File: src/Commands/all/RecallData.ts
import { IStudySet, LineDescriptor } from "@/FlashcardParser/Types/Types";

export class RecallData {
  public static symbol: string = "***";

  public reviewedAt: Date;
  public ease: number;
  public interval: number;
  public learningPhase: boolean;

  constructor(
    lineDescriptor: LineDescriptor,
    studySet: IStudySet,
    args: string
  ) {
    const recallData = parseRecallData(args);
    if (!recallData) {
      console.error(`[RecallData] Failed to parse recall data: "${args}"`);
      return;
    }

    this.reviewedAt = recallData.reviewedAt;
    this.ease = recallData.ease;
    this.interval = recallData.interval;
    this.learningPhase = recallData.learningPhase;

    var p = lineDescriptor.parent;
    if (!p) {
      console.error("[RecallData] No parent");
      return;
    }

    const f = studySet.flashcards.find((f) => f.lineDescriptor === p);
    if (!f) {
      console.error("[RecallData] No flashcard associated");
      return;
    }

    f.reviewedAt = recallData.reviewedAt;
    f.ease = recallData.ease;
    f.interval = recallData.interval;
    f.learningPhase = recallData.learningPhase;

    // Calculate nextReviewAt based on reviewedAt and interval
    if (f.reviewedAt) {
      f.nextReviewAt = new Date(
        f.reviewedAt.getTime() + recallData.interval * 24 * 60 * 60 * 1000
      );
    }

    console.log(
      `[RecallData] Applied to card at line ${p.index}: "${f.text}".` +
      `\n\t- reviewedAt: ${f.reviewedAt?.toISOString()};` +
      `\n\t- nextReviewAt: ${f.nextReviewAt?.toISOString()};` +
      `\n\t- ease: ${f.ease};` +
      `\n\t- interval: ${f.interval};` +
      `\n\t- learningPhase: ${f.learningPhase}`
    );
  }

  public toJson(): object {
    return {
      name: "RecallData",
      vueComponent: null,
      reviewedAt: this.reviewedAt,
      ease: this.ease,
      interval: this.interval,
      learningPhase: this.learningPhase
    };
  }
}

function parseRecallData(recallString: string): {
  reviewedAt: Date | null;
  ease: number;
  interval: number;
  learningPhase: boolean;
} | null {
  try {
    // Expected format: "2025-01-15T10:30:00.000Z, 5, 1.0, false"
    const parts = recallString.split(",").map((part) => part.trim());

    if (parts.length !== 4) {
      console.error(
        `[parser] Invalid recall data format: expected 4 parts, got ${parts.length}`
      );
      return null;
    }

    const reviewedAt = new Date(parts[0]);
    const ease = parseFloat(parts[1]);
    const interval = parseFloat(parts[2]);
    const learningPhase = parts[3].toLowerCase() === "true";

    // Validate parsed data
    if (isNaN(reviewedAt.getTime())) {
      console.error(`[parser] Invalid date: ${parts[0]}`);
      return null;
    }
    if (isNaN(ease) || ease < 1 || ease > 10) {
      console.error(`[parser] Invalid difficulty: ${parts[1]}`);
      return null;
    }
    if (isNaN(interval) || interval < 0) {
      console.error(`[parser] Invalid interval: ${parts[2]}`);
      return null;
    }

    console.log(
      `[parser] Parsed recall data: reviewedAt=${reviewedAt.toISOString()}, ease=${ease}, interval=${interval}, learningPhase=${learningPhase}`
    );

    return {
      reviewedAt,
      ease,
      interval,
      learningPhase,
    };
  } catch (error) {
    console.error(`[parser] Error parsing recall data: ${error}`);
    return null;
  }
}

File: service/flashcard_parser/commands/recall_data.py
from datetime import datetime, timedelta
from ..types import LineDescriptor, StudySet


class RecallData:
    symbol = "***"

    def __init__(self, line_descriptor: LineDescriptor, study_set: StudySet, args: str):
        recall_data = _parse_recall_data(args)
        if not recall_data:
            print(f"[RecallData] Failed to parse recall data: '{args}'")
            return
        self.reviewed_at = recall_data["reviewedAt"]
        self.ease = recall_data["ease"]
        self.interval = recall_data["interval"]
        self.learning_phase = recall_data["learningPhase"]
        parent = line_descriptor.parent
        if parent is None:
            print("[RecallData] No parent")
            return
        f = next((f for f in study_set.flashcards if f.line_descriptor == parent), None)
        if not f:
            print("[RecallData] No flashcard associated")
            return
        f.reviewed_at = self.reviewed_at
        f.ease = self.ease
        f.interval = self.interval
        f.learning_phase = self.learning_phase
        if f.reviewed_at:
            f.next_review_at = f.reviewed_at + timedelta(days=self.interval)

    def to_json(self):
        return {
            "name": "RecallData",
            "reviewedAt": self.reviewed_at,
            "ease": self.ease,
            "interval": self.interval,
            "learningPhase": self.learning_phase,
        }


def _parse_recall_data(recall_string: str):
    try:
        parts = [p.strip() for p in recall_string.split(",")]
        if len(parts) != 4:
            print(f"[parser] Invalid recall data format: expected 4 parts, got {len(parts)}")
            return None
        reviewed_at = datetime.fromisoformat(parts[0])
        ease = float(parts[1])
        interval = float(parts[2])
        learning_phase = parts[3].lower() == "true"
        return {
            "reviewedAt": reviewed_at,
            "ease": ease,
            "interval": interval,
            "learningPhase": learning_phase,
        }
    except Exception as e:
        print(f"[parser] Error parsing recall data: {e}")
        return None
